// @netlify/content-engine/dist/plugin-utils/polyfill-remote-file/jobs/gatsby-worker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.FILE_CDN = FILE_CDN;
exports.IMAGE_CDN = IMAGE_CDN;
var _path = _interopRequireDefault(require("path"));
var _fetchRemoteFile = require("../../../core-utils/fetch-remote-file");
var _cpuCoreCount = require("../../../core-utils/cpu-core-count");
var _fastq = _interopRequireDefault(require("fastq"));
var _transformImages = require("../transform-images");
const queue = (0, _fastq.default)(async function transform(task, cb) {
  try {
    cb(null, await (0, _transformImages.transformImage)(task));
  } catch (e) {
    cb(e);
  }
},
// When inside query workers, we only want to use the current core
process.env.GATSBY_WORKER_POOL_WORKER ? 1 : Math.max(1, (0, _cpuCoreCount.cpuCoreCount)() - 1));
async function FILE_CDN({
  outputDir,
  args: {
    url,
    filename,
    contentDigest,
    httpHeaders
  }
}) {
  const ext = _path.default.extname(filename);
  await (0, _fetchRemoteFile.fetchRemoteFile)({
    directory: outputDir,
    url,
    name: _path.default.basename(filename, ext),
    ext,
    cacheKey: contentDigest,
    excludeDigest: true,
    httpHeaders
  });
}
async function IMAGE_CDN(args) {
  return new Promise((resolve, reject) => {
    queue.push(args, err => {
      if (err) {
        reject(err);
        return;
      }
      resolve();
    });
  });
}
