// @netlify/sdk/dist/integration/connect/model/transformers/scalar/date.js
import { omitByUndefined } from "../../../utils/omit-by-undefined.js";
import { ModelError } from "../../error.js";
import { ScalarBaseTransformer } from "../scalar.js";
export class DateTransformer extends ScalarBaseTransformer {
    parseVisitedInputSingle(input) {
        if (typeof input === `string`)
            return input;
        if (input instanceof Date) {
            return input.toISOString();
        }
        throw new ModelError(`Only strings or Date objects may be passed as input values for Date fields. Found ${typeof input}`);
    }
    toCreateSingleFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "date",
        });
    }
    toCreateListFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "list",
            items: {
                type: "date",
            },
        });
    }
    toCreateSingleFieldValue(input) {
        const value = this.parseVisitedInputSingle(input);
        return {
            type: "date",
            value,
        };
    }
    toCreateListFieldValue(input) {
        return {
            type: `list`,
            items: input?.map?.((listFieldValue) => {
                return {
                    type: "date",
                    value: listFieldValue,
                };
            }),
        };
    }
}
