// @netlify/sdk/dist/integration/connect/model/transformers/scalar/float.js
import { ModelError } from "../../error.js";
import { omitByUndefined } from "../../../utils/omit-by-undefined.js";
import { ScalarBaseTransformer } from "../scalar.js";
export class FloatTransformer extends ScalarBaseTransformer {
    toCreateSingleFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            ...this.modelFieldToSharedCreateNumberProperties(modelField),
            type: "number",
            subtype: "float",
        });
    }
    toCreateListFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "list",
            items: {
                ...this.modelFieldToSharedCreateNumberProperties(modelField),
                type: "number",
                subtype: "float",
            },
        });
    }
    parseVisitedInputSingle(input, field, treeContext) {
        const isNumber = typeof input === `number`;
        if (!input && !field.required && !isNumber) {
            return null;
        }
        if (!isNumber) {
            throw new ModelError(`Input data for field ${field.fieldName} at input path ${treeContext.path.join(`.`)}.${field.fieldName} is not a float but should be. Received ${typeof input}.\n\n${JSON.stringify(input, null, 2)}`);
        }
        return input;
    }
    toCreateSingleFieldValue(input) {
        return {
            type: "number",
            value: input,
        };
    }
    toCreateListFieldValue(input) {
        return {
            type: `list`,
            items: input?.map?.((listFieldValue) => {
                return {
                    type: "number",
                    value: listFieldValue,
                };
            }),
        };
    }
}
