// @netlify/sdk/dist/integration/connect/model/transformers/scalar/html.js
import { StringTransformer } from "./string.js";
import { omitByUndefined } from "../../../utils/omit-by-undefined.js";
export class HtmlTransformer extends StringTransformer {
    toCreateSingleFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "html",
        });
    }
    toCreateListFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "list",
            items: {
                type: "html",
            },
        });
    }
    toCreateSingleFieldValue(input) {
        return {
            type: "html",
            value: input,
        };
    }
    toCreateListFieldValue(input) {
        return {
            type: `list`,
            items: input?.map?.((listFieldValue) => {
                return {
                    type: "html",
                    value: listFieldValue,
                };
            }),
        };
    }
}
