// @netlify/sdk/dist/integration/connect/model/transformers/scalar/json.js
import { omitByUndefined } from "../../../utils/omit-by-undefined.js";
import { ScalarBaseTransformer } from "../scalar.js";
export class JSONTransformer extends ScalarBaseTransformer {
    toCreateSingleFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "json",
        });
    }
    toCreateListFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "list",
            items: {
                type: "json",
            },
        });
    }
    parseVisitedInputSingle(input) {
        return input;
    }
    toCreateSingleFieldValue(input) {
        return {
            type: "json",
            value: input,
        };
    }
    toCreateListFieldValue(input) {
        return {
            type: `list`,
            items: input?.map?.((listFieldValue) => {
                return {
                    type: "json",
                    value: listFieldValue,
                };
            }),
        };
    }
}
