// @netlify/content-engine/dist/bootstrap/load-plugins/load-internal-plugins.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.loadInternalPlugins = loadInternalPlugins;
var _path = require("../../core-utils/path");
var _lodash = _interopRequireDefault(require("lodash.uniqwith"));
var _lodash2 = _interopRequireDefault(require("lodash.isequal"));
var _path2 = _interopRequireDefault(require("path"));
var _processPlugin = require("./process-plugin");
var _createId = require("./utils/create-id");
var _createHash = require("./utils/create-hash");
function loadInternalPlugins(config = {}, rootDir) {
  // Instantiate plugins.
  const plugins = [];

  // Add internal plugins
  const internalPlugins = [`../../internal-plugins/internal-data-bridge`].filter(Boolean);
  internalPlugins.forEach(relPath => {
    const absPath = _path2.default.join(__dirname, relPath);
    plugins.push((0, _processPlugin.processPlugin)(absPath, rootDir));
  });

  // Add plugins from the site config.
  if (config.plugins) {
    config.plugins.forEach(plugin => {
      if (
      // don't process default-site-plugin
      plugin.resolve === `default-site-plugin` ||
      // @ts-ignore
      plugin === `default-site-plugin`) {
        return;
      }
      const processedPlugin = (0, _processPlugin.processPlugin)(plugin, rootDir);
      plugins.push(processedPlugin);
    });
  }

  // Add the site's default "plugin" i.e. engine-node.js file in root of site.
  plugins.push({
    resolve: (0, _path.slash)(process.cwd()),
    id: (0, _createId.createPluginId)(`default-site-plugin`),
    name: `default-site-plugin`,
    version: (0, _createHash.createFileContentHash)(process.cwd(), `gatsby-*`) + (0, _createHash.createFileContentHash)(process.cwd(), `engine-*`),
    pluginOptions: {
      plugins: []
    }
  });
  const uniquePlugins = (0, _lodash.default)(plugins, _lodash2.default);
  return uniquePlugins;
}
